src/mapgen/mapgen.cpp

```diff
@@ -514,15 +514,48 @@ void Mapgen::propagateSunlight(v3s16 nmin, v3s16 nmax, bool propagate_shadow)
 	//printf("propagateSunlight: %dms\n", t.stop());
 }
 
+// findPropagationCutoff
+//
+// Returns the highest y that needs light propagation.
+//
+// Above ground, there is generally a tall tower of open air with only sunlight.
+// There's no need to initiate propagation from these nodes, since they are
+// surrounded by sunlight.
+//
+// This function can skip over these open air nodes using only a single memory
+// access per node. In contrast, spreadLight() requires 7 memory accesses
+// (the node + 6 dirs) to initiate propagation.
+//
+// This speeds up calcLighting by a factor of ~2.6x (measured on X86-64).
+//
+int Mapgen::findPropagationCutoff(const v3s16 &nmin, const v3s16 &nmax)
+{
+	VoxelArea a(nmin, nmax);
+	for (int y = a.MaxEdge.Y; y >= a.MinEdge.Y; y--) {
+		for (int z = a.MinEdge.Z; z <= a.MaxEdge.Z; z++) {
+			u32 i = vm->m_area.index(a.MinEdge.X, y, z);
+			for (int x = a.MinEdge.X; x <= a.MaxEdge.X; x++, i++) {
+				MapNode &n = vm->m_data[i];
+				if (n.getContent() == CONTENT_IGNORE)
+					continue;
+				if (n.param1 == LIGHT_SUN)
```

Comment:
this can only be relied on for nodes which have `param_type == CPT_LIGHT`, I suppose once you find a node that doesn't transmit light you need to run light spread from there.

The second thing I'm wondering is where does param1 get set? Do we know it's going to be correct?

Reply:
Also since `param1` contains two light banks what you're doing here is checking that the node
* is unlit at night
* is maximally lit at day

This sounds correct but should be annotated with a comment because it's easy to miss.

Reply:
n.param1 is set to LIGHT_SUN in `propagateSunlight`, which runs right before spreadLight. Only for nodes that are receiving sunlight. Other nodes remain untouched. At some point they had to have been initialized to 0. (not sure where).

The idea was that if a node doesn't transmit light, there's no way it would be set to LIGHT_SUN.

Unfortunately it appears there's two different kinds of "does not transmit". There's `cf.light_propagates` (which is equivalent to `param_type == CPT_LIGHT`) but also `cf.sunlight_propagates` which is set independently. The latter is used by propagateSunlight.

I suppose this would mean if there's a group of nodes with:

cf.sunlight_propagates =  true
cf.light_propagates = false

Then this optimization could miss some light propagation needed for night light sources. Does that combination make sense / ever happen though? If so, would need to check cf.light_propagates here.

Reply:
[please ignore the deleted comment, I made a logic mistake. Still working out a better understanding]